File: src/main/java/toast/client/modules/Module.java
package toast.client.modules;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.MinecraftClient;
import toast.client.event.EventImpl;
import toast.client.event.EventManager;
import toast.client.event.events.player.EventUpdate;
import toast.client.modules.config.ModuleSettings;
import toast.client.modules.config.Setting;

import java.util.Map;

import static toast.client.ToastClient.CONFIG_MANAGER;

@Environment(EnvType.CLIENT)
public class Module {
    public String name;
    public String description;
    public boolean enabled;
    public int key;
    public Category category;


    public ModuleSettings settings = new ModuleSettings();

    public MinecraftClient mc = MinecraftClient.getInstance();

    public Module(String name, String description, Category category, int key) {
        this.name = name;
        this.description = description;
        this.key = key;
        this.category = category;
        this.settings.addBoolean("Visible", true);
    }

    public String getDescription() {
        return description;
    }

    public boolean isEnabled() {
        return enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
        if (isEnabled()) {
            EventManager.register(this);
            onEnable();
        } else {
            EventManager.unregister(this);
            onDisable();
        }
        CONFIG_MANAGER.writeModules();
    }

    public int getKey() {
        return key;
    }

    public void setKey(int key) {
        this.key = key;
    }

    public String getName() {
        return name;
    }

    public Category getCategory() {
        return category;
    }

    public ModuleSettings getSettings() {
        return settings;
    }

    public void setSettings(Map<String, Setting> newSettings) {
        this.settings.setSettings(newSettings);
    }

    public boolean isMode(String mode) {
        return this.settings.getMode("Mode").equals(mode);
    }

    public String getMode() {
        return this.settings.getMode("Mode");
    }

    public double getDouble(String name) {
        return this.settings.getValue(name);
    }

    public boolean getBool(String name) {
        return this.settings.getBoolean(name);
    }

    public void disable() {
        setEnabled(false);
    }

    public void toggle() {
        setEnabled(!isEnabled());
    }

    public void onEnable() {}

    public void onDisable() {}

    public enum Category {
        PLAYER,
        MOVEMENT,
        RENDER,
        COMBAT,
        MISC
    }

}

File: src/main/java/toast/client/modules/Module.kt
package toast.client.modules

import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.MinecraftClient
import toast.client.ToastClient
import toast.client.modules.config.ModuleSettings

/**
 * Superclass for creating modules
 */
@Environment(EnvType.CLIENT)
open class Module(
    /**
     * The name of the module
     */
    var name: String,

    /**
     * A brief description of what the module does
     */
    var description: String,

    /**
     * The category the module should appear in
     */
    var category: Category,

    /**
     * What key-bind the module should have by default
     */
    var key: Int
) {
    /**
     *
     */

    /**
     * Variable containing the module's configuration options
     */
    var settings: ModuleSettings = ModuleSettings()

    /**
     * Shorthand definition for minecraft instance
     */
    var mc: MinecraftClient = MinecraftClient.getInstance()

    /**
     * Whether or not the module is currently enabled
     */
    var enabled: Boolean = false

    /**
     * Enables or disables a module
     */
    private fun setEnabled(newEnabled: Boolean): Boolean {
        enabled = newEnabled
        if (enabled) {
            try {
                ToastClient.eventBus.register(this@Module)
            } catch (ignored: IllegalArgumentException) {
            }
            onEnable()
        } else {
            try {
                ToastClient.eventBus.unregister(this@Module)
            } catch (ignored: IllegalArgumentException) {
            }
            onDisable()
        }
        return enabled
    }

    /**
     * The current mode of the configuration option called "Mode"
     */
    val mode: String?
        get() = settings.getMode("Mode")

    /**
     * Gets the current value of a configuration option of type Double
     */
    fun getDouble(name: String): Double = settings.getValue(name)!!

    /**
     * Gets the current value of a configuration option of type Boolean
     */
    fun getBool(name: String): Boolean = settings.getBoolean(name)

    /**
     * Disables the module
     */
    fun disable(): Boolean = setEnabled(false)

    /**
     * Enables the module
     */
    fun enable(): Boolean = setEnabled(true)

    /**
     * Toggles the module's enabled state
     */
    fun toggle(): Boolean = setEnabled(!enabled)

    /**
     * Function that gets called whenever the module is enabled
     */
    open fun onEnable() {}

    /**
     * Function that gets called whenever the module is disabled
     */
    open fun onDisable() {}

    /**
     * Enum containing the categories available for modules
     */
    enum class Category {
        /**
         * Modules that aid the player in a none movement or combat related way
         */
        PLAYER,

        /**
         * Modules that help the player move around better
         */
        MOVEMENT,

        /**
         * Modules that affect or add to the game's rendering
         */
        RENDER,

        /**
         * Modules that aid the player in combat
         */
        COMBAT,

        /**
         * Modules that don't fit into any of the other categories
         */
        MISC
    }

    init {
        settings.addBoolean("Visible", true)
    }
}